app/boot_control/_grub.py

```diff
@@ -0,0 +1,607 @@
+        )
+        logger.info(f"{replaced=}")
+        return replaced
+
+    def _update_default_grub(self, patterns):
+        logger.info(f"{patterns=}")
+        lines = open(self._default_grub_file).readlines()
+        patterns_found = {}
+        updated = []
+        for line in lines:
+            found = False
+            for k, v in patterns.items():
+                m = re.match(rf"{k}.*", line)
+                if m:
+                    found = True
+                    updated.append(f"{k}{v}\n")
+                    patterns_found[k] = v
+                    break
+            if not found:
+                updated.append(line)
+
+        logger.info(f"{updated=}")
+        deltas = dict(patterns.items() - patterns_found.items())
+        logger.info(f"{deltas=}")
+        for k, v in deltas.items():
+            updated.append(f"{k}{v}\n")
+
+        with open(self._default_grub_file, "w") as f:
+            f.writelines(updated)
+            f.flush()
+            os.fsync(f.fileno())
+
+    def _grub_reboot_cmd(self, num):
+        cmd = "sync"
+        subprocess.check_output(shlex.split(cmd))
+        cmd = f"grub-reboot {num}"
+        return subprocess.check_output(shlex.split(cmd))
+
+    def _grub_reboot(self):
+        # count grub menu entry number
+        grub_cfg = open(self._grub_cfg_file).read()
+        menus = GrubCfgParser(grub_cfg).parse()
+        self._grub_reboot_cmd(len(menus))
+
+    def _get_cmdline(self):
+        return open("/proc/cmdline").read()
+
+    def _grub_mkconfig_cmd(self, outfile):
+        cmd = f"grub-mkconfig -o {outfile}"
+        subprocess.check_output(shlex.split(cmd))
+        cmd = f"sync {outfile}"
+        subprocess.check_output(shlex.split(cmd))
+
+
+class OtaPartitionFile:
+    """
+    slot: sda2, sda3
+    dev: /dev/sda2, /dev/sda3
+    """
+
+    def __init__(self):
+        self._boot_dir = Path(cfg.BOOT_DIR)  # /boot
+        self._boot_ota_partition_file = Path(
+            cfg.BOOT_OTA_PARTITION_FILE
+        )  # ota-partition
+
+        _ab_detecter = GrubABPartitionDetecter()
+        self.active_root_dev = _ab_detecter.get_active_slot_dev()
+        self.standby_root_dev = _ab_detecter.get_standby_slot_dev()
+        self.active_slot = _ab_detecter.get_active_slot()
+        self.standby_slot = _ab_detecter.get_standby_slot()
+
+        self._grub_control = GrubControl()
+        self._initialize_boot_partition()
+
+    def get_active_root_dev(self):
+        return self.active_root_dev
+
+    def get_active_root_device_name(self):
+        """NOTE: dev name as slot name"""
+        return self.active_slot
+
+    def get_standby_root_dev(self):
+        return self.standby_root_dev
+
+    def get_standby_root_device_name(self):
+        return self.standby_slot
+
+    def get_standby_boot_partition_path(self):
+        slot = self.standby_slot
+        path = self._boot_dir / self._boot_ota_partition_file.with_suffix(f".{slot}")
+        return path
```

Comment:
boot partition is dynamically changed by the boot sequence, but this function seems to return a static value.
Is this behavior OK?

Reply:
Yes, the A/B partition detection and switching mechanism is updated for grub controller now. 

The new active/standby slots and devs detection logic is realized by looking up the partition table, so before a reboot, active/standby slot/dev will always remain the same.

After the local update is finished, before reboot, the switch_boot method will unconditionally point the `ota-partition` symlink to the standby ota-partition folder.

After the first reboot, otaclient will ensure the `ota-partition` symlink points to the active slot ota-partition folder(no matter the update is successful or failed).

Reply:
Sorry, when the reboot fails, due to some kernel boot process, who rollback the boot partition link? automatically rollback is required. grub-reboot can do it but who switch boot partition symlink?

Reply:
Cboot is OK when boot fails 3 times, partition is automatically switch back to the previous one.

Reply:
> Sorry, when the reboot fails, due to some kernel boot process, who rollback the boot partition link? automatically rollback is required. grub-reboot can do it but who switch back boot partition symlink?

I will think it carefully again.